File: Sample/Koudai/Server/release/PyScript/Action/action1402.py
import clr, sys
clr.AddReference('ZyGames.Framework.Common');
clr.AddReference('ZyGames.Framework');
clr.AddReference('ZyGames.Framework.Game');
clr.AddReference('ZyGames.Tianjiexing.Model');
clr.AddReference('ZyGames.Tianjiexing.BLL');
clr.AddReference('ZyGames.Tianjiexing.Lang');
clr.AddReference('ZyGames.Tianjiexing.Component');
from action import *
from System import *
from System.Collections.Generic import *
from ZyGames.Framework.Common.Log import *
from ZyGames.Tianjiexing.Model.ConfigModel import *
from ZyGames.Framework.Common import *
from ZyGames.Tianjiexing.Model import *
from ZyGames.Tianjiexing.BLL import *
from ZyGames.Tianjiexing.Lang import *
from ZyGames.Framework.Game.Cache import *
from ZyGames.Framework.Game.Service import *
from ZyGames.Framework.Game.Runtime import *
from ZyGames.Framework.Common import *
from ZyGames.Framework.Cache.Generic import *
from ZyGames.Tianjiexing.Model.Config import *
from ZyGames.Tianjiexing.Model.Enum import *
from ZyGames.Tianjiexing.BLL.Base import *
from ZyGames.Tianjiexing.Component import *
from ZyGames.Framework.Common.Serialization import *

# 1402_酒馆佣兵列表接口
class UrlParam(HttpParam):
    def __init__(self):
        HttpParam.__init__(self)
        self.recruitList = []
        self.userId = ''
        self.contextUser = None

class ActionResult(DataResult):
    def __init__(self):
        DataResult.__init__(self)        

def getUrlElement(httpGet, parent):
    urlParam = UrlParam();
    urlParam.recruitList = ConfigCacheSet[RecruitRule]().FindAll();

    return urlParam;

def takeAction(urlParam, parent):
    actionResult = ActionResult();
    urlParam.userId = parent.Current.User.PersonalId;
    urlParam.contextUser = parent.Current.User;
    return actionResult;

def buildPacket(writer, urlParam, actionResult):
    
    def getGeneralQuality(_generalQuality):
        genquality = '';
        for quality in _generalQuality:
            genquality += str(MathUtils.ToInt(quality.Quality)) + ",";
        return genquality.TrimEnd(',');

    writer.PushIntoStack(urlParam.contextUser.GoldNum);
    writer.PushIntoStack(urlParam.contextUser.GameCoin);
    writer.PushIntoStack(urlParam.recruitList.Count);
    for rule in urlParam.recruitList:
        mallPrice = MathUtils.ToInt(FestivalHelper.StoreDiscount() * rule.GoldNum);
        isFirst = 0
        rtype = MathUtils.ToEnum[RecruitType](rule.RecruitType)
        if GeneralHelper.IsFirstRecruit(urlParam.contextUser,rtype):
            isFirst = 1
        dsItem = DataStruct();
        dsItem.PushShortIntoStack(MathUtils.ToShort(rule.RecruitType));
        dsItem.PushIntoStack(MathUtils.ToNotNullString(getGeneralQuality(rule.GeneralQuality)));
        dsItem.PushIntoStack(GeneralHelper.SurplusNum(urlParam.userId, rule.FreeNum,MathUtils.ToEnum[RecruitType](rule.RecruitType)));
        dsItem.PushIntoStack(rule.FreeNum);
        dsItem.PushIntoStack(mallPrice);
        dsItem.PushIntoStack(GeneralHelper.UserQueueCodeTime(urlParam.userId, MathUtils.ToEnum[RecruitType](rule.RecruitType)));
        dsItem.PushIntoStack(MathUtils.ToShort(isFirst));
        writer.PushIntoStack(dsItem);
    return True;

File: Sample/Koudai/Server/src/ZyGames.Tianjiexing.BLL/Action/Action1402.cs
/****************************************************************************
Copyright (c) 2013-2015 scutgame.com

http://www.scutgame.com

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
****************************************************************************/
using System;
using System.Collections.Generic;
using ZyGames.Framework.Cache.Generic;
using ZyGames.Framework.Game.Cache;
using ZyGames.Framework.Game.Service;
using ZyGames.Framework.Collection;
using ZyGames.Framework.Common;
using ZyGames.Tianjiexing.BLL.Base;
using ZyGames.Tianjiexing.Lang;
using ZyGames.Tianjiexing.Model;
using ZyGames.Tianjiexing.Model.Config;
using ZyGames.Tianjiexing.Model.ConfigModel;
using ZyGames.Tianjiexing.Model.Enum;


namespace ZyGames.Tianjiexing.BLL.Action
{
    /// <summary>
    /// 1402_酒馆佣兵列表接口
    /// </summary>
    public class Action1402 : BaseAction
    {
        private List<RecruitRule> recruitList = new List<RecruitRule>();

        public Action1402(ZyGames.Framework.Game.Contract.HttpGet httpGet)
            : base(ActionIDDefine.Cst_Action1402, httpGet)
        {

        }

        public override void BuildPacket()
        {
            PushIntoStack(ContextUser.GoldNum);
            PushIntoStack(ContextUser.GameCoin);
            PushIntoStack(recruitList.Count);
            foreach (RecruitRule rule in recruitList)
            {
                DataStruct dsItem = new DataStruct();
                dsItem.PushIntoStack((short)rule.RecruitType);
                dsItem.PushIntoStack(GetGeneralQuality(rule.GeneralQuality).ToNotNullString());
                dsItem.PushIntoStack(GeneralHelper.SurplusNum(ContextUser.UserID, rule.FreeNum, rule.RecruitType.ToEnum<RecruitType>()));
                dsItem.PushIntoStack(rule.FreeNum);
                dsItem.PushIntoStack(rule.GoldNum);
                dsItem.PushIntoStack(GeneralHelper.UserQueueCodeTime(ContextUser.UserID, rule.RecruitType.ToEnum<RecruitType>()));

                this.PushIntoStack(dsItem);
            }

        }

        public override bool GetUrlElement()
        {
            recruitList = new ConfigCacheSet<RecruitRule>().FindAll();
            return true;
        }

        public override bool TakeAction()
        {
            if (true)
            {
                return true;
            }
        }

        public static string GetGeneralQuality(CacheList<RecruitInfo> _generalQuality)
        {
            string genquality = string.Empty;
            foreach (var quality in _generalQuality)
            {
                genquality += quality.Quality.ToInt() + ",";
            }
            return genquality.TrimEnd(',');
        }
    }
}